django/test/testcases.py

```diff
@@ -677,7 +677,7 @@ def _assert_template_used(self, template_name, template_names, msg_prefix, count
             template_name in template_names,
             msg_prefix + "Template '%s' was not a template used to render"
             " the response. Actual template(s) used: %s"
-            % (template_name, ", ".join(template_names)),
+            % (template_name, str(template_names)),
```

Comment:
This is backward incompatible, we should rather modify `_AssertTemplateUsedContext.on_template_render()`.

Reply:
backward incompatible as in it changes the style of the output?
We thought it's beneficial to switch to a standard list representation here since that is easy to understand and people see that a template without a name was used.
If that is not an option, would you prefer to filter None types out completely or replace them in `_AssertTemplateUsedContext.on_template_render()` with a default string to indicate the usage of an unnamed template?

Reply:
> backward incompatible as in it changes the style of the output?

Yes, and not compatible with ` assertTemplateUsed()` used normally, not as a context manager.

>  If that is not an option, would you prefer to filter None types out completely or replace them in _AssertTemplateUsedContext.on_template_render() with a default string to indicate the usage of an unnamed template?

We should do the same as in:

https://github.com/django/django/blob/e41f9f94505f690525f16af8bd10e62e22bc0207/django/test/testcases.py#L670

Reply:
When doing something like this, it will show that No templates were used if just one unnamed template is actually used, is that fine?

https://github.com/django/django/pull/17237/commits/25557a1b16a498b776eca4c501820877c164f5c3
See this for a new test leading to this issue.

Reply:
```
diff --git a/django/test/testcases.py b/django/test/testcases.py
index ac1996d97b..1a700d8689 100644
--- a/django/test/testcases.py
+++ b/django/test/testcases.py
@@ -116,18 +116,16 @@ class _AssertTemplateUsedContext:
         self.count = count
 
         self.rendered_templates = []
-        self.rendered_template_names = []
         self.context = ContextList()
 
     def on_template_render(self, sender, signal, template, context, **kwargs):
         self.rendered_templates.append(template)
-        self.rendered_template_names.append(template.name)
         self.context.append(copy(context))
 
     def test(self):
         self.test_case._assert_template_used(
             self.template_name,
-            self.rendered_template_names,
+            [t.name for t in self.rendered_templates if t.name is not None],
             self.msg_prefix,
             self.count,
         )
@@ -146,7 +144,7 @@ class _AssertTemplateUsedContext:
 class _AssertTemplateNotUsedContext(_AssertTemplateUsedContext):
     def test(self):
         self.test_case.assertFalse(
-            self.template_name in self.rendered_template_names,
+            self.template_name in [t.name for t in self.rendered_templates if t.name is not None],
             f"{self.msg_prefix}Template '{self.template_name}' was used "
             f"unexpectedly in rendering the response",
         )
```

A patch like this would lead to this error message for the newly added test above: 'No templates used to render the response'

Reply:
> what do you think about the patch suggested above compared to the original one?

Looks better, thanks.